File: EntityModels.Postgresql/TerritoryEvent.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace EntityModels.Postgresql;

[Table("TerritoryEvent")]
public abstract partial class TerritoryEvent
{
    [Key] public string Name { get; set; } = null!;
    [StringLength(50)] public string Type { get; set; } = null!;
    [StringLength(1000)] public string Description { get; set; } = null!;
    [StringLength(1000)] public string? IconPath { get; set; }
    public int Capacity { get; set; }
    public int DurationSeconds { get; set; }
    [StringLength(50)] public string InstrumentType { get; set; } = null!;
    public int InstrumentRequiredLevel { get; set; }
    public int InstrumentExpectedLevel { get; set; } //  (x2 faster for level above and 4x slower for level below)

    public ICollection<TerritoryEventItem> TerritoryEventItems { get; set; } = null!;
    public ICollection<TerritoryEventProportion> TerritoryEventProportions { get; set; } = null!;
}

File: client/src/models/region/TerritoryEvent.ts
import { TerritoryEventItem } from "./TerritoryEventItem";
import { ObjectWithProportion as owp } from "../ObjectWithProportion";

export class TerritoryEvent extends owp<TerritoryEventItem[]> {
  constructor(obj: TerritoryEventItem[],
              name: string,
              type: string,
              description: string,
              capacity: number,
              duration: number,
              instrumentType: string,
              instrumentRequiredLevel: number,
              instrumentExpectedLevel: number) {
    super(obj, 0, name, description);
    this.obj = obj;
    this.name = name;
    this.type = type;
    this.capacity = capacity;
    this.duration = duration;
    this.instrumentType = instrumentType;
    this.instrumentRequiredLevel = instrumentRequiredLevel;
    this.instrumentExpectedLevel = instrumentExpectedLevel;
  }

  declare obj: TerritoryEventItem[];
  name: string;
  type: string;
  capacity: number;
  duration: number;
  instrumentType: string;
  instrumentRequiredLevel: number;
  instrumentExpectedLevel: number;
}